pkg/expr/convert_to_long.go

```diff
@@ -0,0 +1,314 @@
+	timeIndexMap := make(map[time.Time]int, len(times))
+	for i, t := range times {
+		timeIndexMap[t] = i
+	}
+
+	// Populate output fields
+	for _, frame := range frames {
+		var timeField *data.Field
+		for _, field := range frame.Fields {
+			if field.Type() == data.FieldTypeTime {
+				timeField = field
+				break
+			}
+		}
+
+		if timeField == nil {
+			return nil, fmt.Errorf("no time field found in frame")
+		}
+
+		for _, field := range frame.Fields {
+			if field.Type().Numeric() {
+				for i := 0; i < field.Len(); i++ {
+					t := timeField.At(i).(time.Time)
+					val, err := field.FloatAt(i)
+					if err != nil {
+						val = 0 // Default value for missing data
+					}
+					idx := timeIndexMap[t]
+					if outputField, exists := outputNumericFields[field.Name]; exists {
+						outputField.Set(idx, val)
+					}
+
+					// Add labels for the numeric field
+					for key, value := range field.Labels {
+						if outputField, exists := outputLabelFields[key]; exists {
+							outputField.Set(idx, value)
+						}
+					}
+				}
+			}
+		}
+	}
+
+	// Build the output frame
+	outputFields := []*data.Field{timeField}
+	for _, field := range outputNumericFields {
+		outputFields = append(outputFields, field)
+	}
+	for _, field := range outputLabelFields {
+		outputFields = append(outputFields, field)
+	}
+	outputFrame := data.NewFrame("time_series_long", outputFields...)
+
+	// Set metadata
+	if outputFrame.Meta == nil {
+		outputFrame.Meta = &data.FrameMeta{}
+	}
+	outputFrame.Meta.Type = data.FrameTypeTimeSeriesLong
+
+	return data.Frames{outputFrame}, nil
+}
+
+func convertTimeSeriesWideToTimeSeriesLong(frames data.Frames) (data.Frames, error) {
+	// Wide should only be one frame
+	if len(frames) != 1 {
+		return nil, fmt.Errorf("expected exactly one frame for wide format, but got %d", len(frames))
+	}
+	inputFrame := frames[0]
+	longFrame, err := data.WideToLong(inputFrame)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert wide time series to long timeseries for sql expression: %w", err)
+	}
+	return data.Frames{longFrame}, nil
+}
+
+func getToLongConversionFunc(inputType data.FrameType) func(data.Frames) (data.Frames, error) {
+	switch inputType {
+	case data.FrameTypeNumericMulti:
+		return convertNumericMultiToNumericLong
+	case data.FrameTypeNumericWide:
+		return convertNumericWideToNumericLong
+	case data.FrameTypeTimeSeriesMulti:
+		return convertTimeSeriesMultiToTimeSeriesLong
+	case data.FrameTypeTimeSeriesWide:
+		return convertTimeSeriesWideToTimeSeriesLong
+	default:
+		return convertErr
+	}
+}
+
+func convertErr(_ data.Frames) (data.Frames, error) {
+	return nil, fmt.Errorf("unsupported input type for SQL expression")
+}
+
+func supportedToLongConversion(inputType data.FrameType) bool {
+	switch inputType {
+	case data.FrameTypeNumericMulti:
+		return true
+	case data.FrameTypeNumericWide:
+		return true
+	case data.FrameTypeTimeSeriesMulti:
+		return true
+	case data.FrameTypeTimeSeriesWide:
+		return true
```

Comment:
I have a slight preference for this personally:
```suggestion
	switch inputType {
	case data.FrameTypeNumericMulti, data.FrameTypeNumericWide, data.FrameTypeTimeSeriesMulti, data.FrameTypeTimeSeriesWide:
		return true
```
Not sure what you think ❓

Reply:
Those would include the long formats, which we don't want here.